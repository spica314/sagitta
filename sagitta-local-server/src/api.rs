use std::path::PathBuf;

use actix_web::{web, App, HttpServer};
use sagitta_common::clock::Clock;

use crate::api_state::ApiState;

use self::v1::sync::v1_sync;

pub mod v1;

pub struct ServerConfig {
    pub clock: Clock,
    pub port: u16,
    pub local_system_workspace_base_path: PathBuf,
    pub remote_api_base_url: String,
}

pub async fn run_local_api_server(config: ServerConfig) {
    let state = ApiState::new(
        config.clock.clone(),
        config.local_system_workspace_base_path.clone(),
        &config.remote_api_base_url,
    )
    .await;

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .service(v1_sync)
    })
    .bind(("0.0.0.0", config.port))
    .unwrap()
    .run()
    .await
    .unwrap();
}
