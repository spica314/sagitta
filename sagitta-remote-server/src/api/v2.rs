pub mod commit;
pub mod create_workspace;
pub mod get_attr;
pub mod get_file_blob_id;
pub mod get_workspace_id_from_name;
pub mod get_workspaces;
pub mod read_blob;
pub mod read_dir;
pub mod sync_files_with_workspace;
pub mod write_blob;
