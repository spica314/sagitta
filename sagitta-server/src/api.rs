use std::path::PathBuf;

use actix_web::{web, App, HttpServer};
use sagitta_common::clock::Clock;
use sagitta_objects::{SagittaTreeObject, SagittaTreeObjectFile};

use crate::state::ApiState;

use self::{
    blob::read::blob_read,
    trunk::get_head::trunk_get_head,
    workspace::{create::workspace_create, get_head::workspace_get_head, list::workspace_list},
};

pub mod blob;
pub mod trunk;
pub mod workspace;

pub struct ServerConfig {
    pub base_path: PathBuf,
    pub surreal_uri: String,
    pub is_main: bool,
    pub clock: Clock,
    pub port: u16,
}

pub async fn run_server(config: ServerConfig) {
    let state = ApiState::new(
        config.base_path.clone(),
        config.clock.clone(),
        &config.surreal_uri,
        config.is_main,
    )
    .await;

    // root (dir1)
    // - hello.txt (file1)
    // - hello_dir (dir2)
    //     - hello2.txt (file2)

    let file1_blob_id = state
        .remote_system_workspace_manager
        .save_object(None, b"Hello, world!\n".as_slice())
        .unwrap();
    let file1 = SagittaTreeObject::File(SagittaTreeObjectFile {
        blob_id: file1_blob_id,
        size: 14,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o644,
    });
    let file1_id = state
        .remote_system_workspace_manager
        .save_tree(None, &file1)
        .unwrap();

    let file2_blob_id = state
        .remote_system_workspace_manager
        .save_object(None, b"Hello, world!!\n".as_slice())
        .unwrap();
    let file2 = SagittaTreeObject::File(SagittaTreeObjectFile {
        blob_id: file2_blob_id,
        size: 15,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o644,
    });
    let file2_id = state
        .remote_system_workspace_manager
        .save_tree(None, &file2)
        .unwrap();

    let tree2 = SagittaTreeObject::Dir(sagitta_objects::SagittaTreeObjectDir {
        items: vec![("hello2.txt".to_string(), file2_id)],
        size: 4096,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o755,
    });
    let tree2_id = state
        .remote_system_workspace_manager
        .save_tree(None, &tree2)
        .unwrap();

    let tree1 = SagittaTreeObject::Dir(sagitta_objects::SagittaTreeObjectDir {
        items: vec![
            ("hello.txt".to_string(), file1_id),
            ("hello_dir".to_string(), tree2_id),
        ],
        size: 4096,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o755,
    });
    let tree1_id = state
        .remote_system_workspace_manager
        .save_tree(None, &tree1)
        .unwrap();

    let commit = sagitta_objects::SagittaCommitObject {
        tree_id: tree1_id,
        parent_commit_id: None,
        message: "Initial commit".to_string(),
    };
    let commit_id = state
        .remote_system_workspace_manager
        .save_commit(None, &commit)
        .unwrap();
    state
        .remote_system_workspace_manager
        .set_trunk_head(&commit_id)
        .await
        .unwrap();

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .service(blob_read)
            .service(trunk_get_head)
            .service(workspace_create)
            .service(workspace_list)
            .service(workspace_get_head)
    })
    .bind(("0.0.0.0", config.port))
    .unwrap()
    .run()
    .await
    .unwrap();
}
