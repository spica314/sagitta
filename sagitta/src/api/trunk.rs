pub mod get_head;
