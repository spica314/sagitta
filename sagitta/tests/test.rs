use std::time::SystemTime;

use sagitta::api::ServerConfig;
use sagitta_api_schema::{
    blob::read::{BlobReadRequest, BlobReadResponse},
    trunk::get_head::{TrunkGetHeadRequest, TrunkGetHeadResponse},
};
use sagitta_objects::{SagittaCommitObject, SagittaTreeObject};
use tempfile::tempdir;
use tokio::runtime::Builder;

#[test]
fn test_1() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    let config = ServerConfig {
        base_path: tempdir.as_ref().to_path_buf(),
        clock: sagitta::tools::Clock::new_with_fixed_time(fixed_system_time),
    };

    runtime.spawn(async {
        sagitta::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let head_id_res: TrunkGetHeadResponse = ureq::post("http://localhost:8081/trunk/get-head")
        .send_json(TrunkGetHeadRequest {})
        .unwrap()
        .into_json()
        .unwrap();
    let head_id = head_id_res.id;
    insta::assert_debug_snapshot!(head_id);

    let commit_res: BlobReadResponse = ureq::post("http://localhost:8081/blob/read")
        .send_json(BlobReadRequest { id: head_id })
        .unwrap()
        .into_json()
        .unwrap();
    let commit: SagittaCommitObject = serde_cbor::from_reader(commit_res.blob.as_slice()).unwrap();
    insta::assert_debug_snapshot!(commit);

    let dir_res: BlobReadResponse = ureq::post("http://localhost:8081/blob/read")
        .send_json(BlobReadRequest { id: commit.tree_id })
        .unwrap()
        .into_json()
        .unwrap();
    let dir: SagittaTreeObject = serde_cbor::from_reader(dir_res.blob.as_slice()).unwrap();
    insta::assert_debug_snapshot!(dir);

    let SagittaTreeObject::Dir(dir) = dir else {
        panic!()
    };
    let mut xs = vec![];
    for item in &dir.items {
        let file_res: BlobReadResponse = ureq::post("http://localhost:8081/blob/read")
            .send_json(BlobReadRequest { id: item.clone() })
            .unwrap()
            .into_json()
            .unwrap();
        let file: SagittaTreeObject = serde_cbor::from_reader(file_res.blob.as_slice()).unwrap();
        let SagittaTreeObject::File(file) = file else {
            panic!()
        };

        let blob_res: BlobReadResponse = ureq::post("http://localhost:8081/blob/read")
            .send_json(BlobReadRequest { id: file.blob_id })
            .unwrap()
            .into_json()
            .unwrap();
        let blob = std::str::from_utf8(blob_res.blob.as_slice()).unwrap();
        insta::assert_debug_snapshot!(blob);

        xs.push((file.name.clone(), blob.to_string()));
    }
    insta::assert_debug_snapshot!(xs);
}
