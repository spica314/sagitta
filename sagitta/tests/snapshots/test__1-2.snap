---
source: sagitta/tests/test.rs
expression: commit
---
SagittaCommitObject {
    tree_id: ObjectId {
        id: "ae2811c34e7a9364d679dcde6efc47d8056ea2927835acc8af936008e8a9a21c",
    },
    parent_commit_id: None,
    message: "Initial commit",
}
